multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -11,205 +11,452 @@
 {
-	dataBarier();
-	/* Disable Transfer Complete interrupt, disable channel */
-	*(channel_base + ccr) &= ~((1 << 1) | 0x1);
-	dataBarier();
+	int dma = per->dma, channel = per->channel[dir];
+	char reqmap = per->reqmap;
+
+	libdma_configureChannel(dma, channel, dir, priority, paddr, msize, psize, minc, pinc, reqmap, cond);
+
+	return 0;
 }
 
 
-static int _has_channel_finished(const void *maddr, volatile unsigned int *channel_base)
+static int libdma_hasChannelFinished(const void *maddr, volatile unsigned int *channel_base)
 {
-	return (maddr != NULL && *(channel_base + cndtr) > 0);
+	return (maddr != NULL) && (*(channel_base + cndtr) == 0);
 }
 
 
-static int _transfer(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
+static int libdma_transferHelper(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
 {
 	/* Empirically chosen value to avoid mutex+cond overhead for short transactions */
 	int use_interrupts = len > 24;
+	int interrupts_flags = (use_interrupts == 0) ? 0 : DMA_TCIE_FLAG;
+	volatile int rx_done = 0;
 
-	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + 5 * rx_channel;
-	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + 5 * tx_channel;
+	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + (5 * rx_channel);
+	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + (5 * tx_channel);
 
-	if (rx_maddr != NULL)
-		_prepare_transfer(rx_channel_base, rx_maddr, len, use_interrupts);
+	/* Only one request may be issued on one channel at one time. */
+	if ((dma_transfers[dma][rx_channel].type != dma_transferNull) ||
+		(dma_transfers[dma][tx_channel].type != dma_transferNull)) {
+		return -EINVAL;
+	}
 
-	if (tx_maddr != NULL)
-		/* When doing rw transfer, avoid unnecessary interrupt handling and condSignal()
-		by waiting only for RX transfer completion, ignoring TX */
-		_prepare_transfer(tx_channel_base, (void *)tx_maddr, len, use_interrupts && rx_maddr == NULL);
+	dma_transfers[dma][rx_channel].type = dma_transferOnce;
+	dma_transfers[dma][rx_channel].once.done_flag = &rx_done;
+	dma_transfers[dma][rx_channel].once.channel_base = rx_channel_base;
+	libdma_prepareTransfer(rx_channel_base, rx_maddr, len, interrupts_flags);
 
-	if (use_interrupts) {
+	dma_transfers[dma][tx_channel].type = dma_transferOnce;
+	/* When doing rw transfer, avoid unnecessary interrupt handling and condSignal()
+	by waiting only for RX transfer completion, ignoring TX */
+	libdma_prepareTransfer(tx_channel_base, (void *)tx_maddr, len, 0);
+
+	if (use_interrupts != 0) {
 		mutexLock(dma_common[dma].irqLock);
-		while (_has_channel_finished(rx_maddr, rx_channel_base) || _has_channel_finished(tx_maddr, tx_channel_base))
+		while ((rx_done == 0) || (libdma_hasChannelFinished(tx_maddr, tx_channel_base) == 0)) {
 			condWait(dma_common[dma].cond, dma_common[dma].irqLock, 1);
```

Comment:
What about this timeout? I think it can be set to 0.

BTW, `time_t timeout` is a rather unfortunate parameter user-wise - neither the type nor the name tells which unit of time to use. `time_t` has no consistent meaning, because sometimes it is used for seconds (e.g. `time()`) and sometimes for microseconds (e.g. `condWait()`, `gettime()`). Don't bother for now, this is more of a note for the future, because a rational solution should be sought in all the relevant modules/repos.

Reply:
I informed @agkaminski  about it he decided that it will be further investigated but not covered by this change

Reply:
I suspect timeout was meant to be `1000 * 1000` - timeout of 1 us will cause instant timeout and busy waiting.